sdks/java/expansion-service/src/main/java/org/apache/beam/sdk/expansion/service/ExpansionServiceSchemaTransformProvider.java

```diff
@@ -0,0 +1,174 @@
+import org.apache.beam.vendor.grpc.v1p48p1.com.google.protobuf.InvalidProtocolBufferException;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+@SuppressWarnings({"rawtypes"})
+public class ExpansionServiceSchemaTransformProvider implements TransformProvider {
+
+  static final String DEFAULT_INPUT_TAG = "INPUT";
+
+  private Map<String, org.apache.beam.sdk.schemas.transforms.SchemaTransformProvider>
+      schemaTransformProviders = new HashMap<>();
+  private static @Nullable ExpansionServiceSchemaTransformProvider transformProvider = null;
+
+  private ExpansionServiceSchemaTransformProvider() {
+    try {
+      for (org.apache.beam.sdk.schemas.transforms.SchemaTransformProvider schemaTransformProvider :
+          ServiceLoader.load(
+              org.apache.beam.sdk.schemas.transforms.SchemaTransformProvider.class)) {
+        if (schemaTransformProviders.containsKey(schemaTransformProvider.identifier())) {
+          throw new IllegalArgumentException(
+              "Found multiple SchemaTransformProvider implementations with the same identifier "
+                  + schemaTransformProvider.identifier());
+        }
+        schemaTransformProviders.put(schemaTransformProvider.identifier(), schemaTransformProvider);
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e.getMessage());
+    }
+  }
+
+  public static ExpansionServiceSchemaTransformProvider of() {
+    if (transformProvider == null) {
+      transformProvider = new ExpansionServiceSchemaTransformProvider();
+    }
+
+    return transformProvider;
+  }
+
+  static class RowTransform extends PTransform {
+
+    private PTransform<PCollectionRowTuple, PCollectionRowTuple> rowTuplePTransform;
+
+    public RowTransform(PTransform<PCollectionRowTuple, PCollectionRowTuple> rowTuplePTransform) {
+      this.rowTuplePTransform = rowTuplePTransform;
+    }
+
+    @Override
+    public POutput expand(PInput input) {
+      PCollectionRowTuple inputRowTuple;
+
+      if (input instanceof PCollectionRowTuple) {
+        inputRowTuple = (PCollectionRowTuple) input;
+      } else if (input instanceof PCollection) {
+        inputRowTuple = PCollectionRowTuple.of(DEFAULT_INPUT_TAG, (PCollection) input);
+      } else if (input instanceof PBegin) {
+        inputRowTuple = PCollectionRowTuple.empty(input.getPipeline());
+      } else if (input instanceof PCollectionTuple) {
+        inputRowTuple = PCollectionRowTuple.empty(input.getPipeline());
+        PCollectionTuple inputTuple = (PCollectionTuple) input;
+        for (TupleTag<?> tag : inputTuple.getAll().keySet()) {
+          inputRowTuple = inputRowTuple.and(tag.getId(), (PCollection<Row>) inputTuple.get(tag));
+        }
+      } else {
+        throw new RuntimeException(String.format("Unsupported input type: %s", input));
+      }
+      PCollectionRowTuple output = inputRowTuple.apply(this.rowTuplePTransform);
+
+      if (output.getAll().size() > 1) {
```

Comment:
Is there any advantage to doing this? (In particular, the name of the single output gets lost.)